pyomo/core/base/set.py

```diff
@@ -1655,27 +1715,30 @@ class OrderedSetData(_OrderedSetMixin, FiniteSetData):
 
     def __init__(self, component):
         self._values = {}
-        self._ordered_values = []
+        self._ordered_values = None
         FiniteSetData.__init__(self, component=component)
 
     def _iter_impl(self):
         """
         Return an iterator for the set.
         """
-        return iter(self._ordered_values)
+        return iter(self._values)
 
     def __reversed__(self):
-        return reversed(self._ordered_values)
+        return reversed(self._values)
 
-    def _add_impl(self, value):
-        self._values[value] = len(self._values)
-        self._ordered_values.append(value)
+    def _update_impl(self, values):
+        for val in values:
+            # Note that we reset _ordered_values within the loop because
+            # of an old example where the initializer rule makes
+            # reference to values previously inserted into the Set
+            # (which triggered the creation of the _ordered_values)
+            self._ordered_values = None
```

Comment:
The offending example is:
```python
def U_init(model, z):
    if z == 6:
        return Set.End
    if z == 1:
        return 1
    else:
        return model.U[z - 1] * z

model.U = Set(ordered=True, initialize=U_init)
```
The issue is that when `z == 2` the `__getitem__` causes the `_ordered_values` list to be created with a single entry.  If we don't clear `_ordered_values`, then it will stay a 1-element list, and when `z == 3`, looking for `model.U[2]` looks up `_ordered_values[1]`; running off the end of the list.

Reply:
Ewwww, this is horrifying! But okay, makes sense.